File: frontend/radvis/src/app/viewer/weitere-kartenebenen/services/weitere-kartenebenen.guard.ts
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */
import { inject } from '@angular/core';
import { CanActivateFn, Router } from '@angular/router';
import { NotifyUserService } from 'src/app/shared/services/notify-user.service';
import { VIEWER_ROUTE } from 'src/app/viewer/viewer-shared/models/viewer-routes';
import { WeitereKartenebenenRoutingService } from 'src/app/viewer/weitere-kartenebenen/services/weitere-kartenebenen-routing.service';

export const weitereKartenebenenGuard: CanActivateFn = () => {
  const service: WeitereKartenebenenRoutingService = inject(WeitereKartenebenenRoutingService);
  const router: Router = inject(Router);
  const notifyUserService: NotifyUserService = inject(NotifyUserService);
  if (service.currentFeature !== null) {
    return true;
  }
  notifyUserService.inform('Gewähltes Feature nicht mehr verfügbar.');
  return router.createUrlTree([VIEWER_ROUTE]);
};

File: backend/src/main/java/de/wps/radvis/backend/weitereKartenebenen/schnittstelle/WeitereKartenebenenGuard.java
/*
 * Copyright (c) 2023 WPS - Workplace Solutions GmbH
 *
 * Licensed under the EUPL, Version 1.2 or as soon they will be approved by the European Commission - subsequent versions of the EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package de.wps.radvis.backend.weitereKartenebenen.schnittstelle;

import java.nio.file.AccessDeniedException;
import java.util.List;

import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.RequestBody;

import de.wps.radvis.backend.benutzer.domain.BenutzerResolver;
import de.wps.radvis.backend.benutzer.domain.entity.Benutzer;
import de.wps.radvis.backend.benutzer.domain.valueObject.Recht;
import jakarta.validation.Valid;

public class WeitereKartenebenenGuard {
	private final BenutzerResolver benutzerResolver;

	public WeitereKartenebenenGuard(BenutzerResolver benutzerResolver) {
		this.benutzerResolver = benutzerResolver;
	}

	public void save(Authentication authentication,
		@RequestBody List<@Valid SaveWeitereKartenebeneCommand> commands) throws AccessDeniedException {
		Benutzer benutzer = benutzerResolver.fromAuthentication(authentication);

		if (benutzer.hatRecht(Recht.WEITERE_KARTENEBENEN_ALS_DEFAULT_FESTLEGEN)) {
			return;
		}

		if (commands.stream().filter(c -> c.getId() == null).anyMatch(c -> c.isDefaultLayer())) {
			throw new AccessDeniedException("Sie sind nicht berechtigt, Layer für alle Nutzer hinzuzufügen.");
		}
	}
}
